compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -4601,7 +4631,103 @@ abstract class GenJSCode extends plugins.PluginComponent
           patchedBody,
           allActualCaptures)
 
-      JSFunctionToScala(closure, params.size)
+      // Wrap the closure in the appropriate box for the SAM type
+      val funSym = originalFunction.tpe.typeSymbolDirect
+      if (isFunctionSymbol(funSym)) {
+        /* This is a scala.FunctionN. We use the existing AnonFunctionN
+         * wrapper.
+         */
+        JSFunctionToScala(closure, params.size)
+      } else {
+        /* This is an arbitrary SAM type (can only happen in 2.12).
+         * We have to synthesize a class like LambdaMetaFactory would do on
+         * the JVM.
+         */
+        val sam = originalFunction.attachments.get[SAMFunctionCompat].fold[Symbol] {
+          abort(s"Cannot find the SAMFunction attachment on $originalFunction at $pos")
+        } {
+          _.sam
+        }
+
+        val samWrapperClassName = synthesizeSAMWrapper(funSym, sam)
+        js.New(jstpe.ClassType(samWrapperClassName), js.Ident("init___O"),
+            List(closure))
+      }
+    }
+
+    private def synthesizeSAMWrapper(funSym: Symbol, sam: Symbol)(
+        implicit pos: Position): String = {
+      val intfName = encodeClassFullName(funSym)
+
+      val suffix = {
+        val newCount = generatedSAMWrapperCount.value + 1
+        generatedSAMWrapperCount.value = newCount
+        "$$Lambda$" + newCount // LambdaMetaFactory names classes like this
+      }
+      val generatedClassName = encodeClassFullName(currentClassSym) + suffix
+
+      val genClassType = jstpe.ClassType(generatedClassName)
+
+      // val f$1: Any
+      val funFieldIdent = js.Ident("f$1", Some("f"))
+      val funFieldDef = js.FieldDef(funFieldIdent, jstpe.AnyType, mutable = false)
+
+      // def this(f: Any) = { this.f$1 = f; super() }
+      val ctorDef = {
+        val fParamDef = js.ParamDef(js.Ident("f"), jstpe.AnyType,
+            mutable = false, rest = false)
+        js.MethodDef(static = false, js.Ident("init___O"), List(fParamDef),
+            jstpe.NoType,
+            js.Block(List(
+                js.Assign(
+                    js.Select(js.This()(genClassType), funFieldIdent)(jstpe.AnyType),
+                    fParamDef.ref),
```

Comment:
Consider aligning `funFieldIdent` and `fParamDef`: Use either both times just `f` or both times `fun`.